src/Common/test/Common.Test/Configuration/StringExtensionsTest.cs

```diff
@@ -0,0 +1,29 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+using FluentAssertions;
+using Steeltoe.Common.Configuration;
+using Xunit;
+
+namespace Steeltoe.Common.Test.Configuration;
+
+public sealed class StringExtensionsTest
+{
+    [Theory]
+    [InlineData("foobar", "foobar")]
+    [InlineData("foo.bar", "foo:bar")]
+    [InlineData("foo__bar", "foo:bar")]
+    [InlineData("foo__bar_", "foo:bar_")]
+    [InlineData("foo__bar__", "foo:bar__")]
+    [InlineData("foo[bar]", "foo[bar]")]
+    [InlineData("foobar[1234]", "foobar:1234")]
+    [InlineData("foobar[1234][5678]", "foobar:1234:5678")]
+    [InlineData("foobar[1234][5678]barbar", "foobar[1234][5678]barbar")]
+    [InlineData("a.b.foobar[1234][5678].barfoo.boo[123]", "a:b:foobar:1234:5678:barfoo:boo:123")]
+    [InlineData(@"a.b\.foobar", "a:b.foobar")]
```

Comment:
These should contain three segments (to ensure it works multiple times) of different lengths (to ensure we're at the right offsets), ie: `one.four.seven` instead of `foo.bar`.

I'm wondering if the following cases are correct:
- `foo__bar__`: Looking into the built-in .NET implementation, it doesn't preserve the trailing __, so why should we?
- `foobar[1234][5678]barbar` is this a legal key, or should it be: `foobar[1234][5678].barbar`?
- `a.b\.foobar` would be better to cover both escape cases, ie: `[InlineData(@"one\.four\\.seven", @"one.four\:seven")]`

Reply:
> I'm wondering if the following cases are correct:
> 
> * `foo__bar__`: Looking into the built-in .NET implementation, it doesn't preserve the trailing __, so why should we?
> * `foobar[1234][5678]barbar` is this a legal key, or should it be: `foobar[1234][5678].barbar`?
> * `a.b\.foobar` would be better to cover both escape cases, ie: `[InlineData(@"one\.four\\.seven", @"one.four\:seven")]`

The last two are pre-existing test cases from the ConfigServer code. I don't mind adding the additional escaping, and I think the point of the `[1234][5678]` case is that it should be left alone. For the trailing underscore cases, I didn't think they would be valid, so intended to leave keys with those patterns as-is.

Reply:
Related https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-Configuration-Binding#map-based-binding and https://github.com/SteeltoeOSS/Steeltoe/commit/42df149c7d11ae53e7b3d85157a66985ff98f18e.